### PR TITLE
Homework 4 version 1

### PR DESCRIPTION
**Pull request by n3jr4**

### CODE REVIEW

File: src/org/nejrasm/zadaca4oop/task1/Duck.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package org.nejrasm.zadaca4oop.task1;
+
+public class Duck implements Flight, Swim {
+    @Override
+    public String fly() {
+        return new Pegasus().fly();
```
Comment (basictarik) [inline src/org/nejrasm/zadaca4oop/task1/Duck.java]: :D this does go a bit too far. Separation of concern is preferred. Also you had to create a new Pegasus object just to reuse it's method. What if Pegasus actually had arguments for it's constructor, how would you provide them?

One way to do what you wanted was following:
1.
```
public class FlyWithWIngs implements Flight {
    public String fly() {
        return "Flies with wings";
    }
}
```
Give Duck and pegasus a flying behaviour:
```
public class Duck implements Flight {
    Flight flyingBehaviour;
    public Duck() {
        flyingBehaviour = new FlyWithWings();
    }
    public String fly() {
        flyingBehaviour.fly();
    }
}
```

In this particular case it is not even required for duck or pegasus to implement flight. Anyway this was not the goal of the exercise.

I appreciate that you try to reuse code, that is very good, but just make the same method for simplicity now.

File: src/org/nejrasm/zadaca4oop/task2/Rectangle.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package org.nejrasm.zadaca4oop.task2;
+
+public class Rectangle extends Shape {
+    private final double a;
+    private final double b;
+
+    protected Rectangle(final String colour, final double a, final double b) {
```
Comment (basictarik) [inline src/org/nejrasm/zadaca4oop/task2/Rectangle.java]: You force protected too much. Yes you hide the ctor, which is generally good, but this class is not usable outside of the package  `org.nejrasm.zadaca4oop.task2;`. That is not good. Don't rely on the design choice of protected to be visible in the same package, rather make this public. It's important for attributes to be private not that much for methods.

File: src/org/nejrasm/zadaca4oop/task2/Shape.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+        this.colour = colour;
+    }
+
+    protected abstract double getArea();
+
+    protected abstract double getCircumference();
+
+    @Override
+    public final String toString() {
+        return "Colour: {" + this.colour + "}, Area: {" + this.getArea() + "}, Circumference: {" + this.getCircumference() + "}";
```
Comment (basictarik) [inline src/org/nejrasm/zadaca4oop/task2/Shape.java:16]: very nice, elegant solution to the toString problem

**Review by basictarik — APPROVED** [review]
Very well done.

Please read the comments below. Those aren't big mistakes, just try keep things simple.

File: src/org/nejrasm/zadaca4oop/task2/Rectangle.java (hunk 2 above)
Comment (n3jr4) [reply src/org/nejrasm/zadaca4oop/task2/Rectangle.java]: Noted.

File: src/org/nejrasm/zadaca4oop/task1/Duck.java (hunk 1 above)
Comment (n3jr4) [reply src/org/nejrasm/zadaca4oop/task1/Duck.java]: Noted :)

**Pull request merged**